File: anthropic/typescript/src/services/messages/stream.ts
import type { Message, ContentBlock, Usage, StopReason, TextBlock } from './types.js';

// Stream event types
export type MessageStreamEvent =
  | MessageStartEvent
  | ContentBlockStartEvent
  | ContentBlockDeltaEvent
  | ContentBlockStopEvent
  | MessageDeltaEvent
  | MessageStopEvent
  | PingEvent
  | ErrorEvent;

export interface MessageStartEvent {
  type: 'message_start';
  message: Partial<Message>;
}

export interface ContentBlockStartEvent {
  type: 'content_block_start';
  index: number;
  content_block: ContentBlock;
}

export interface ContentBlockDeltaEvent {
  type: 'content_block_delta';
  index: number;
  delta: ContentDelta;
}

export interface ContentDelta {
  type: 'text_delta' | 'input_json_delta' | 'thinking_delta';
  text?: string;
  partial_json?: string;
  thinking?: string;
}

export interface ContentBlockStopEvent {
  type: 'content_block_stop';
  index: number;
}

export interface MessageDeltaEvent {
  type: 'message_delta';
  delta: { stop_reason?: StopReason; stop_sequence?: string };
  usage?: Partial<Usage>;
}

export interface MessageStopEvent {
  type: 'message_stop';
}

export interface PingEvent {
  type: 'ping';
}

export interface ErrorEvent {
  type: 'error';
  error: { type: string; message: string };
}

// MessageStream class
export class MessageStream implements AsyncIterable<MessageStreamEvent> {
  private reader: ReadableStreamDefaultReader<Uint8Array>;
  private buffer: string = '';
  private isDone: boolean = false;

  constructor(stream: ReadableStream<Uint8Array>) {
    this.reader = stream.getReader();
  }

  async *[Symbol.asyncIterator](): AsyncIterator<MessageStreamEvent> {
    const decoder = new TextDecoder();

    while (!this.isDone) {
      const { done, value } = await this.reader.read();

      if (done) {
        this.isDone = true;
        break;
      }

      this.buffer += decoder.decode(value, { stream: true });

      // Parse SSE events from buffer
      const events = this.parseEvents();
      for (const event of events) {
        yield event;
        if (event.type === 'message_stop' || event.type === 'error') {
          this.isDone = true;
        }
      }
    }
  }

  private parseEvents(): MessageStreamEvent[] {
    const events: MessageStreamEvent[] = [];
    const lines = this.buffer.split('\n');

    // Keep the last incomplete line in the buffer
    this.buffer = lines.pop() ?? '';

    let currentEvent: { event?: string; data?: string } = {};

    for (const line of lines) {
      if (line.startsWith('event: ')) {
        currentEvent.event = line.slice(7).trim();
      } else if (line.startsWith('data: ')) {
        currentEvent.data = line.slice(6).trim();
      } else if (line === '' && currentEvent.data) {
        // Empty line signals end of event
        try {
          const eventType = currentEvent.event || 'message';
          const eventData = JSON.parse(currentEvent.data);

          // Map event type to our event interface
          const event = this.mapEvent(eventType, eventData);
          if (event) {
            events.push(event);
          }
        } catch (error) {
          // Ignore parse errors for malformed events
          console.error('Failed to parse SSE event:', error);
        }
        currentEvent = {};
      }
    }

    return events;
  }

  private mapEvent(eventType: string, data: any): MessageStreamEvent | null {
    switch (eventType) {
      case 'message_start':
        return {
          type: 'message_start',
          message: data.message,
        };
      case 'content_block_start':
        return {
          type: 'content_block_start',
          index: data.index,
          content_block: data.content_block,
        };
      case 'content_block_delta':
        return {
          type: 'content_block_delta',
          index: data.index,
          delta: data.delta,
        };
      case 'content_block_stop':
        return {
          type: 'content_block_stop',
          index: data.index,
        };
      case 'message_delta':
        return {
          type: 'message_delta',
          delta: data.delta,
          usage: data.usage,
        };
      case 'message_stop':
        return { type: 'message_stop' };
      case 'ping':
        return { type: 'ping' };
      case 'error':
        return {
          type: 'error',
          error: data.error,
        };
      default:
        return null;
    }
  }

  async collect(): Promise<Message> {
    const accumulator = new MessageStreamAccumulator();

    for await (const event of this) {
      accumulator.add(event);

      if (event.type === 'error') {
        throw new Error(`Stream error: ${event.error.message}`);
      }
    }

    return accumulator.toMessage();
  }

  async cancel(): Promise<void> {
    await this.reader.cancel();
    this.isDone = true;
  }
}

// Stream accumulator for convenience
export class MessageStreamAccumulator {
  private message: Partial<Message> = {
    content: [],
    usage: {
      input_tokens: 0,
      output_tokens: 0,
    },
  };
  private contentBlocks: ContentBlock[] = [];
  private currentBlockText: Map<number, string> = new Map();
  private currentBlockJson: Map<number, string> = new Map();
  private currentThinking: Map<number, string> = new Map();

  add(event: MessageStreamEvent): void {
    switch (event.type) {
      case 'message_start':
        this.message = {
          ...this.message,
          ...event.message,
          content: [],
        };
        break;

      case 'content_block_start':
        this.contentBlocks[event.index] = event.content_block;
        if (event.content_block.type === 'text') {
          this.currentBlockText.set(event.index, event.content_block.text);
        } else if (event.content_block.type === 'thinking') {
          this.currentThinking.set(event.index, event.content_block.thinking);
        } else if (event.content_block.type === 'tool_use') {
          this.currentBlockJson.set(event.index, JSON.stringify(event.content_block.input));
        }
        break;

      case 'content_block_delta':
        this.handleDelta(event.index, event.delta);
        break;

      case 'content_block_stop':
        this.finalizeBlock(event.index);
        break;

      case 'message_delta':
        if (event.delta.stop_reason) {
          this.message.stop_reason = event.delta.stop_reason;
        }
        if (event.delta.stop_sequence) {
          this.message.stop_sequence = event.delta.stop_sequence;
        }
        if (event.usage) {
          this.message.usage = {
            ...this.message.usage!,
            ...event.usage,
          };
        }
        break;

      case 'message_stop':
        // Finalize message
        this.message.content = this.contentBlocks;
        break;

      case 'ping':
        // Ignore ping events
        break;

      case 'error':
        // Error will be handled by the caller
        break;
    }
  }

  private handleDelta(index: number, delta: ContentDelta): void {
    switch (delta.type) {
      case 'text_delta':
        if (delta.text) {
          const current = this.currentBlockText.get(index) || '';
          this.currentBlockText.set(index, current + delta.text);
        }
        break;

      case 'input_json_delta':
        if (delta.partial_json) {
          const current = this.currentBlockJson.get(index) || '';
          this.currentBlockJson.set(index, current + delta.partial_json);
        }
        break;

      case 'thinking_delta':
        if (delta.thinking) {
          const current = this.currentThinking.get(index) || '';
          this.currentThinking.set(index, current + delta.thinking);
        }
        break;
    }
  }

  private finalizeBlock(index: number): void {
    const block = this.contentBlocks[index];
    if (!block) return;

    if (block.type === 'text') {
      const text = this.currentBlockText.get(index);
      if (text !== undefined) {
        this.contentBlocks[index] = { type: 'text', text };
      }
    } else if (block.type === 'thinking') {
      const thinking = this.currentThinking.get(index);
      if (thinking !== undefined) {
        this.contentBlocks[index] = { type: 'thinking', thinking };
      }
    } else if (block.type === 'tool_use') {
      const json = this.currentBlockJson.get(index);
      if (json !== undefined) {
        try {
          const input = JSON.parse(json);
          this.contentBlocks[index] = {
            ...block,
            input,
          };
        } catch {
          // Keep the original block if JSON parsing fails
        }
      }
    }
  }

  get content(): string {
    return this.contentBlocks
      .filter((b): b is TextBlock => b && b.type === 'text')
      .map(b => b.text)
      .join('');
  }

  toMessage(): Message {
    if (!this.message.id || !this.message.model) {
      throw new Error('Incomplete message: missing required fields');
    }

    return {
      id: this.message.id,
      type: 'message',
      role: 'assistant',
      content: this.contentBlocks.filter(b => b !== undefined),
      model: this.message.model,
      stop_reason: this.message.stop_reason ?? null,
      stop_sequence: this.message.stop_sequence ?? null,
      usage: this.message.usage ?? {
        input_tokens: 0,
        output_tokens: 0,
      },
    };
  }
}

File: anthropic/rust/src/services/messages/stream.rs
//! Streaming support for the Messages API

use super::types::{
    ContentBlock, ContentDelta, Message, MessageDelta, PartialMessage, Role, StopReason, Usage,
};
use crate::error::{AnthropicError, StreamError};
use bytes::Bytes;
use futures::{Stream, StreamExt};
use pin_project_lite::pin_project;
use serde::Deserialize;
use std::pin::Pin;
use std::task::{Context, Poll};

/// Stream events from the Messages API
#[derive(Debug, Clone, PartialEq)]
pub enum MessageStreamEvent {
    /// Message started
    MessageStart {
        message: PartialMessage,
    },
    /// Content block started
    ContentBlockStart {
        index: usize,
        content_block: ContentBlock,
    },
    /// Content block delta (incremental update)
    ContentBlockDelta {
        index: usize,
        delta: ContentDelta,
    },
    /// Content block stopped
    ContentBlockStop {
        index: usize,
    },
    /// Message delta (metadata update)
    MessageDelta {
        delta: MessageDelta,
        usage: Usage,
    },
    /// Message stopped
    MessageStop,
    /// Ping event (keepalive)
    Ping,
    /// Error event
    Error {
        error: StreamError,
    },
}

/// Internal SSE event structure
#[derive(Debug, Deserialize)]
struct SseEvent {
    #[serde(rename = "type")]
    event_type: String,
    #[serde(flatten)]
    data: serde_json::Value,
}

/// Message start event
#[derive(Debug, Deserialize)]
struct MessageStartEvent {
    message: PartialMessage,
}

/// Content block start event
#[derive(Debug, Deserialize)]
struct ContentBlockStartEvent {
    index: usize,
    content_block: ContentBlock,
}

/// Content block delta event
#[derive(Debug, Deserialize)]
struct ContentBlockDeltaEvent {
    index: usize,
    delta: ContentDelta,
}

/// Content block stop event
#[derive(Debug, Deserialize)]
struct ContentBlockStopEvent {
    index: usize,
}

/// Message delta event
#[derive(Debug, Deserialize)]
struct MessageDeltaEvent {
    delta: MessageDelta,
    usage: Usage,
}

pin_project! {
    /// Stream of message events
    pub struct MessageStream {
        #[pin]
        inner: Box<dyn Stream<Item = Result<Bytes, AnthropicError>> + Send + Unpin>,
        buffer: String,
        is_done: bool,
        // State for accumulating the final message
        current_message: Option<PartialMessage>,
        content_blocks: Vec<ContentBlock>,
        accumulated_text: Vec<String>,
    }
}

impl MessageStream {
    /// Create a new message stream
    pub fn new(
        inner: Box<dyn Stream<Item = Result<Bytes, AnthropicError>> + Send + Unpin>,
    ) -> Self {
        Self {
            inner,
            buffer: String::new(),
            is_done: false,
            current_message: None,
            content_blocks: Vec::new(),
            accumulated_text: Vec::new(),
        }
    }

    /// Collect the stream into a complete message
    pub async fn collect(mut self) -> Result<Message, AnthropicError> {
        let mut message: Option<PartialMessage> = None;
        let mut content_blocks: Vec<ContentBlock> = Vec::new();
        let mut accumulated_text: Vec<String> = Vec::new();
        let mut stop_reason: Option<StopReason> = None;
        let mut stop_sequence: Option<String> = None;
        let mut final_usage: Option<Usage> = None;

        while let Some(event) = self.next().await {
            match event? {
                MessageStreamEvent::MessageStart { message: msg } => {
                    message = Some(msg);
                }
                MessageStreamEvent::ContentBlockStart {
                    index,
                    content_block,
                } => {
                    while content_blocks.len() <= index {
                        content_blocks.push(ContentBlock::Text {
                            text: String::new(),
                            cache_control: None,
                        });
                        accumulated_text.push(String::new());
                    }
                    content_blocks[index] = content_block;
                }
                MessageStreamEvent::ContentBlockDelta { index, delta } => {
                    if let Some(text) = delta.text {
                        while accumulated_text.len() <= index {
                            accumulated_text.push(String::new());
                        }
                        accumulated_text[index].push_str(&text);
                    }
                }
                MessageStreamEvent::MessageDelta { delta, usage } => {
                    stop_reason = delta.stop_reason;
                    stop_sequence = delta.stop_sequence;
                    final_usage = Some(usage);
                }
                MessageStreamEvent::MessageStop => {
                    break;
                }
                MessageStreamEvent::Error { error } => {
                    return Err(AnthropicError::Stream(error.to_string()));
                }
                _ => {}
            }
        }

        // Update content blocks with accumulated text
        for (i, text) in accumulated_text.iter().enumerate() {
            if i < content_blocks.len() {
                if let ContentBlock::Text { cache_control, .. } = &content_blocks[i] {
                    content_blocks[i] = ContentBlock::Text {
                        text: text.clone(),
                        cache_control: cache_control.clone(),
                    };
                }
            }
        }

        let partial_message = message
            .ok_or_else(|| AnthropicError::Stream("No message start event received".to_string()))?;

        let usage = final_usage.unwrap_or(partial_message.usage);

        Ok(Message {
            id: partial_message.id,
            message_type: partial_message.message_type,
            role: partial_message.role,
            content: content_blocks,
            model: partial_message.model,
            stop_reason,
            stop_sequence,
            usage,
        })
    }

    /// Parse an SSE event from a line
    fn parse_sse_line(&mut self, line: &str) -> Option<Result<MessageStreamEvent, AnthropicError>> {
        if line.is_empty() {
            return None;
        }

        if !line.starts_with("data: ") {
            return None;
        }

        let data = &line[6..]; // Remove "data: " prefix

        if data == "[DONE]" {
            return Some(Ok(MessageStreamEvent::MessageStop));
        }

        // Parse the JSON event
        match serde_json::from_str::<SseEvent>(data) {
            Ok(event) => self.parse_event(event),
            Err(e) => Some(Err(AnthropicError::Stream(format!(
                "Failed to parse SSE event: {}",
                e
            )))),
        }
    }

    /// Parse a typed event
    fn parse_event(&self, event: SseEvent) -> Option<Result<MessageStreamEvent, AnthropicError>> {
        match event.event_type.as_str() {
            "message_start" => {
                match serde_json::from_value::<MessageStartEvent>(event.data) {
                    Ok(data) => Some(Ok(MessageStreamEvent::MessageStart {
                        message: data.message,
                    })),
                    Err(e) => Some(Err(AnthropicError::Stream(format!(
                        "Failed to parse message_start: {}",
                        e
                    )))),
                }
            }
            "content_block_start" => {
                match serde_json::from_value::<ContentBlockStartEvent>(event.data) {
                    Ok(data) => Some(Ok(MessageStreamEvent::ContentBlockStart {
                        index: data.index,
                        content_block: data.content_block,
                    })),
                    Err(e) => Some(Err(AnthropicError::Stream(format!(
                        "Failed to parse content_block_start: {}",
                        e
                    )))),
                }
            }
            "content_block_delta" => {
                match serde_json::from_value::<ContentBlockDeltaEvent>(event.data) {
                    Ok(data) => Some(Ok(MessageStreamEvent::ContentBlockDelta {
                        index: data.index,
                        delta: data.delta,
                    })),
                    Err(e) => Some(Err(AnthropicError::Stream(format!(
                        "Failed to parse content_block_delta: {}",
                        e
                    )))),
                }
            }
            "content_block_stop" => {
                match serde_json::from_value::<ContentBlockStopEvent>(event.data) {
                    Ok(data) => Some(Ok(MessageStreamEvent::ContentBlockStop {
                        index: data.index,
                    })),
                    Err(e) => Some(Err(AnthropicError::Stream(format!(
                        "Failed to parse content_block_stop: {}",
                        e
                    )))),
                }
            }
            "message_delta" => {
                match serde_json::from_value::<MessageDeltaEvent>(event.data) {
                    Ok(data) => Some(Ok(MessageStreamEvent::MessageDelta {
                        delta: data.delta,
                        usage: data.usage,
                    })),
                    Err(e) => Some(Err(AnthropicError::Stream(format!(
                        "Failed to parse message_delta: {}",
                        e
                    )))),
                }
            }
            "message_stop" => Some(Ok(MessageStreamEvent::MessageStop)),
            "ping" => Some(Ok(MessageStreamEvent::Ping)),
            "error" => Some(Err(AnthropicError::Stream(format!(
                "Error event: {:?}",
                event.data
            )))),
            _ => None, // Ignore unknown events
        }
    }
}

impl Stream for MessageStream {
    type Item = Result<MessageStreamEvent, AnthropicError>;

    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        let mut this = self.project();

        if *this.is_done {
            return Poll::Ready(None);
        }

        // Poll the inner stream for more data
        match this.inner.as_mut().poll_next(cx) {
            Poll::Ready(Some(Ok(bytes))) => {
                // Add bytes to buffer
                let text = String::from_utf8_lossy(&bytes);
                this.buffer.push_str(&text);

                // Process complete lines
                if let Some(newline_pos) = this.buffer.find('\n') {
                    let line = this.buffer[..newline_pos].to_string();
                    *this.buffer = this.buffer[newline_pos + 1..].to_string();

                    if let Some(event) = this.parse_sse_line(&line) {
                        return Poll::Ready(Some(event));
                    }

                    // If we didn't get an event, poll again
                    cx.waker().wake_by_ref();
                    Poll::Pending
                } else {
                    // Need more data
                    cx.waker().wake_by_ref();
                    Poll::Pending
                }
            }
            Poll::Ready(Some(Err(e))) => {
                *this.is_done = true;
                Poll::Ready(Some(Err(e)))
            }
            Poll::Ready(None) => {
                *this.is_done = true;
                Poll::Ready(None)
            }
            Poll::Pending => Poll::Pending,
        }
    }
}
